[WOOMOB-382] Update Coupons Fetching/Loading Business Logic

malinajirka opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #WOOMOB-382
<!-- Id number of the GitHub issue this PR addresses. -->

**Do not merge label - https://github.com/woocommerce/woocommerce-android/pull/14015 needs to be merged first.**

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR updates the behavior of the coupons list in the POS.
- Show cached data when available.
- Show PTR loading indicator when fetching from remote and cached data shown to indicate a request is in progress.
- Prevent users from starting another PTR when the previous one is in progres.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

1. Clean coupons cache
2. Open POS
3. Switch to Coupons tab
4. Notice "fullscreen" loading is shown (since cache is empty)
5. Notice the data are shown when the request finishes
------------
1. Exit POS
2. Open POS
3. Switch to Coupons tab
4. Notice cached data are shown and PTR indicator is shown since the remote request is in progress
5. Notice the coupons cache is replaced with the fetched data
-----------
1. Pull to refresh
2. Notice a loading indicator is shown
3. Notice you can't pull to refresh again until the request finishes
4. Notice the loading indicator disappears when the request finishes and the data are replaced with the fetched data
5. 
### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->

I've tested the above.


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/fc64bd43-444a-498b-9fc8-b59bd1b4d24f


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

malinajirka committed:
Show cached coupons when fetching first page

malinajirka committed:
Add tests for WooPosCouponsListViewStateManager

dangermattic commented:
<!--
  1 Error: This PR is tagged with <code>s...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

malinajirka committed:
Fix detekt

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/coupons/WooPosCouponsListViewStateManager.kt:
```diff
@@ -59,17 +63,22 @@ class WooPosCouponsListViewStateManager @Inject constructor(
         }
     }
 
+    private suspend fun WooPosCouponsListViewStateManager.showCachedData(
```
Why it needs to be an extension of `WooPosCouponsListViewStateManager`?

kidinov replied:
Also, the naming is a bit confusing. The method itself doesn't show anything and can create it from any coupons, not necessary from cached. Maybe simply `createContentViewState`?

malinajirka replied:
Agree, I'll update this.

malinajirka replied:
Fixed in https://github.com/woocommerce/woocommerce-android/pull/14019/commits/c9d9efb2cb89c9e548842859061aef03955df66d

kidinov reviewed (commented):
A few things I noticed:

* The PTR is shown when it's not triggered by the user in a few cases

https://github.com/user-attachments/assets/9d863874-d32d-4478-a7e4-01dfe0cea95d

https://github.com/user-attachments/assets/4967ef5e-56cf-40d0-a2f8-de2a58f30d31


* On the empty state the PTR is not working properly as it's closed before the remote request comes back

https://github.com/user-attachments/assets/1eb624d1-6a89-4eb0-bbac-2c5d622be708


* Not related to the PR - empty state is at the top

malinajirka commented:
Thanks for the review @kidinov! All the issues you mentioned don't feel like issues to me - I mean I implemented the behavior like this intentionally (except the vertical alignment). Wdyt @samiuelson, how does the behavior feel to you? Long story short, the app displays `Loading more` indicator in progress when we show cache data and a remote request is in progress. If we are not showing cached data and remote request is in progress we display the shimmering effect rows.

> The PTR is shown when it's not triggered by the user in a few cases

I added it intentionally, isn't it desired to show we are loading something? Or how do we show that the remote request is in progress?

> Not related to the PR - empty state is at the top

This is already fixed in one of the other open PRs.

> On the empty state the PTR is not working properly as it's closed before the remote request comes back

Again, I believe this is desired - we are showing "fullscreen" loading in this state and showing two loading approaches on top of each other feels off, doesn't it?

kidinov commented:
@malinajirka  👋

Aside from that, it feels off to me (this may not be important at all 😀); more important is that it's at least partially against the guidelines and not consistent with the behavior of the products and variations in the POS

> The refresh indicator appears only in conjunction with a refresh gesture or action. Syncing does not display a refresh indicator.

> The refresh indicator remains visible until the refresh activity completes and any new content is visible, or the user navigates away from the refreshing content.

https://m2.material.io/design/platform-guidance/android-swipe-to-refresh.html#usage

malinajirka commented:
Thanks for sharing these references!

> On the empty state the PTR is not working properly as it's closed before the remote request comes back

👍  I'll update this one to continue showing the indicator until the request finishes to match material 3 guidelines.

**Update: This is fixed now.**

> The refresh indicator appears only in conjunction with a refresh gesture or action. Syncing does not display a refresh indicator.

I believe tapping on retry button doesn't go against the guideline. It's still a refresh gesture or action. Wdyt?

> more important is that it's at least partially against the guidelines

Not showing any indicator when action is in progress and then out of no-where replacing the displayed data or even worse replacing them with network error is very likely  against the guidelines as well.  (btw we show the PTR indicator on order list and product lists as well when remote request is in progress - I'm not saying that's a reason we should match designs in the POS, just saying that it's not something new in our app and not something users or developers reported AFAIK)

It seems that guidelines say we should be showing a horizontal indicator at the top of the list -https://m3.material.io/components/progress-indicators/guidelines#0804b871-234a-44d3-be91-0ec85e161ee5. I think we can consider updating the lists to match this. (AFAIK this doesn't affect product list since it's loaded on POS load.)

However, regarding `and not consistent with the behavior of the products and variations in the POS`: those list are different in more aspects - not observable, data fetched on POS load, coupons are stored in persistent storage, ... . We could consider bringing them to consistency, however, not showing any indicator when load is in progress is imo a bug not something we should try to implement across the POS.

malinajirka committed:
Rename method from showCachedData to createContentViewState

malinajirka committed:
Replace when with if statement

malinajirka committed:
Continue showing PTR when refreshing empty state

malinajirka committed:
Show PTR only when manually triggered

malinajirka committed:
Fix detekt

malinajirka committed:
Show loading more row when fetching first page

malinajirka commented:
I plan to test this a bit more, but I've updated the behavior as you suggested. Thanks for sharing the examples of the gmail/whatsapp on Slack - those truly don't show any loading indicator when they are updating the cache. I still think we should either show it or we should continue showing the cached data when remote request fails, but we iterate on that later.

I'll re-request review after I test all the changes.

malinajirka committed:
Remove commented out code

malinajirka commented:
I've rested it and it seems it all works now as you suggested. Would appreciate if you could re-review it, thanks!

kidinov commented:
@malinajirka 👋

> I believe tapping on retry button doesn't go against the guideline. It's still a refresh gesture or action. Wdyt?

I don't see that they say anything specific about this case, and they allow it to be displayed when data is requested to be refreshed, e.g. this video:

https://github.com/user-attachments/assets/078f6fa8-d2db-4ebf-b337-174b3bc5ee95

`Refreshing content generated by an app bar action.`

But IMO, it's better to display a skeleton, as the primary goal of the PTR is to be a **swipe gesture** that combines the gesture with the indicator. In this case, it's just the indicator, without the gesture. Another point is the consistency with the rest of the lists - I believe they don't have PTR enabled for the error states at all.

That said, if the chosen architecture makes it difficult to implement this, I think it's ok to leave it as it is, since this error state, which hopefully nobody will see anyway

> Not showing any indicator when action is in progress and then out of nowhere replacing the displayed data or even worse replacing them with network error is very likely against the guidelines as well. 

I believe this was already discussed on Slack. We verified that numerous major apps all display cached data without any loading indicators for new data fetching, updating the UI only when the data arrives. In my opinion, this approach is sensible, particularly for our situation where the data does not change frequently. Therefore, in most cases, merchants will begin using the app without seeing for any background processes we may be running

> It seems that guidelines say we should be showing a horizontal indicator at the top of the list -https://m3.material.io/components/progress-indicators/guidelines#0804b871-234a-44d3-be91-0ec85e161ee5. I think we can consider updating the lists to match this. (AFAIK this doesn't affect product list since it's loaded on POS load.)

I don't see anywhere that it says that. Both video and text say that:

> Progress indicators inform users about the status of ongoing processes, such as **loading an app**, **submitting a form**, or **saving updates**.

But nothing like this is happening here. We just sync the data. In the video examples, they show that they have no data to display; therefore, there is a loading indicator. By the way, I completely don't think that we need to introduce a third loading indicator on top of the skeleton and PTR here.

> However, regarding and not consistent with the behavior of the products and variations in the POS: those list are different in more aspects - not observable, data fetched on POS load, coupons are stored in persistent storage, ... .

The lists themselves are the same from the user's point of view, and I understand that's the technical approach you've decided on, but it shouldn't truly affect the overall user experience. Overall, if the architecture has limitations on such small things like when to display the PTR or when to show a skeleton, I believe it's possibly not the best choice for this particular case.

malinajirka commented:
@kidinov Thanks for the reply! 

I'm a bit confused though - I updated the app to behave as you suggested, do you still see some issues with the current approach?

malinajirka committed:
Merge branch 'trunk' into issue/woomob-382-woo-poscoupons-show-cached-coupons-before-remote-fetch

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/coupons/WooPosCouponsViewModel.kt
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/items/coupons/WooPosCouponsListViewStateManagerTest.kt

kidinov commented:
@malinajirka 👋

Sorry, I didn't test the code yet, just answered you 😀

I missed this comment 😮‍💨 

> I've rested it and it seems it all works now as you suggested. Would appreciate if you could re-review it, thanks!



Testing it now!

kidinov reviewed (approved):
Looks nice, thanks!

The only things I noticed:

* Instead of the skeleton, we show the cached data when there is a retry on the error screen
* I think when it's PTR on the empty screen, we should just show PTR. No need to show two loading indicators, but I guess this is more like NP
* Not sure about that, but it looks like both empty and error screens have a larger margin at the top than at the bottom. I didn't measure that, just how it looks to my eye, so maybe it's not correct
* I think I mentioned this already before, but it looks like we use color for the empty state icon that has not have enough of contrast with the background

https://github.com/user-attachments/assets/4f764ec8-fe2a-4319-8d80-dfcfcd1dfa3e


https://github.com/user-attachments/assets/bde51bef-283a-4571-b4b6-cb30d7f21370

malinajirka committed:
Show loading instead of cached data on retry

malinajirka committed:
Add tests for retry action

malinajirka committed:
Fix detekt

malinajirka commented:
Thanks for the re-review @kidinov !

> Instead of the skeleton, we show the cached data when there is a retry on the error screen


I originally didn't update this since it was consistent with the product list, but looking at it it's not great UX. I've fixed this state on the coupon list - it's still happening on the product list though. When the internet is down, the error arrives so quickly that it's not visible, but on poor internet connection the cached data are shown. Do you think it's worth updating there? If so, could you please update it?

| No Data in Cache                                                                        | Data in cache                                                                       |
|------------------------------------------------------------------------------------|------------------------------------------------------------------------------------|
| <video src="https://github.com/user-attachments/assets/a0415e45-b374-423b-9485-a96bc7e72ea5" width="350px" controls> | <video src="https://github.com/user-attachments/assets/251e240f-002b-4d88-bfc7-b4ee3c1b65fc" width="350px" controls> |

> I think when it's PTR on the empty screen, we should just show PTR. No need to show two loading indicators, but I guess this is more like NP

Make sense, however, this feels like a state that would make the code more complicated => more costly to maintain => the value for the user is IMO zero, the value for the company is therefor likely negative.

> I think I mentioned this already before, but it looks like we use color for the empty state icon that has not have enough of contrast with the background

A couple week ago Wagner mentioned he plans to polish all the error/empty screen and we should ignore them for now.

> Not sure about that, but it looks like both empty and error screens have a larger margin at the top than at the bottom. I didn't measure that, just how it looks to my eye, so maybe it's not correct

It seems to be center correctly (340px from top and bottom) - assuming we are centering it within the list not center of the screen.

| Top                                                                                  | Bottom                                                                               |
|-------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|
| <img src="https://github.com/user-attachments/assets/760cd6cd-e431-4d2c-82a1-0d44689cf0fc" alt="Screenshot 2025-05-08 at 14 14 39" width="300px"> | <img src="https://github.com/user-attachments/assets/b9a30a61-d121-4f45-ace3-fa71b24bd681" alt="Screenshot 2025-05-08 at 14 14 29" width="300px"> |

kidinov commented:
@malinajirka 

> I originally didn't update this since it was consistent with the product list, but looking at it it's not great UX. I've fixed this state on the coupon list - it's still happening on the product list though. When the internet is down, the error arrives so quickly that it's not visible, but on poor internet connection the cached data are shown. Do you think it's worth updating there? If so, could you please update it?

Apologies, but I'm somewhat unclear about that. Update what exactly? We have two ways to address remote request failures while utilizing a local cache:

* We display an error screen when a remote request fails because we want to prioritize the accuracy of the remote data, even if we have cached data. Therefore, the cache is invalidated if the remote fails. In this case, when we click retry, we show a skeleton to indicate that we are loading data. **That's how the product list is currently implemented.**
* We don't display the error screen when a remote request fails and we have cached data, as we show cached data in this case; thus, we clearly don't have a retry button. However, we do show the error screen when a remote request fails and we lack cached data; therefore, we cannot display cached data when the retry button is clicked, as we don't have it.

So in both cases, the retry button cannot show cached data, only a skeleton. Here, it seems that a mix of both was implemented. If we have cached data, we should not show the error screen at all

> Make sense, however, this feels like a state that would make the code more complicated => more costly to maintain => the value for the user is IMO zero, the value for the company is therefor likely negative.

I won't say that it's zero. For me, small things like that create an impression of a good product where people did their best. However, I agree that it's probably not the case for most people, and the impact is really really low.

As per implementation, it largely depends on the selected architectural approach. In the products, this is managed by those lines of code:

```
            _viewState.value = if (withPullToRefresh) {
                buildProductsReloadingState()
            } else {
                WooPosProductsViewState.Loading()
            }
```

Which, in my opinion, isn't very complicated. However, if you find it complicated in your situation, it's perfectly acceptable to leave it as is

> A couple week ago Wagner mentioned he plans to polish all the error/empty screen and we should ignore them for now.

👍

> It seems to be center correctly (340px from top and bottom) - assuming we are centering it within the list not center of the screen.

👍

kidinov reviewed (approved):
LGTM. Please go ahead with mering if you feel it's good to go

malinajirka commented:
> Apologies, but I'm somewhat unclear about that. Update what exactly? We have two ways to address remote request failures while utilizing a local cache:

> We display an error screen when a remote request fails because we want to prioritize the accuracy of the remote data, even if we have cached data. Therefore, the cache is invalidated if the remote fails. In this case, when we click retry, we show a skeleton to indicate that we are loading data. That's how the product list is currently implemented.
> We don't display the error screen when a remote request fails and we have cached data, as we show cached data in this case; thus, we clearly don't have a retry button. However, we do show the error screen when a remote request fails and we lack cached data; therefore, we cannot display cached data when the retry button is clicked, as we don't have it.

Interesting, now I'm even more confused 🤔 . How do we decide whether to invalidate the cache after a request failure? What do we display on the UI when a request fails and we continue showing the cached data? (I mean, I'm aware only about the `load more` failure).

Btw, I have never seen an app which would clear cached data upon a request-failure. It feels like we are trying too be too smart, ending up with worse UX. I think the user can decide whether they are fine continue using the product if the refresh failed. Especially considering the order creation is always remote, so all the calculations are guaranteed to be correct.

> Which, in my opinion, isn't very complicated. However, if you find it complicated in your situation, it's perfectly acceptable to leave it as is

Yes, if you don't look at the big picture, I agree handling such cases is simple. It's similar with VM size - at the beginning one might feel like we can put everything there, since it's just 80 lines of code. However, if we continue applying this we end up with 500+ lines of code and the way back is very expensive. 

In this case, it might be one `if` statement, but if we handle 15 such cases, we have 15 `if` statements in a better case, in the worse case even more since the states affect each other. Every single `if` statement needs to be understood by whoever wants to change the class => it is often difficult to understand them when they fix random edge cases which often don't matter.

kidinov commented:
> Interesting, now I'm even more confused 🤔 . How do we decide whether to invalidate the cache after a request failure? What do we display on the UI when a request fails and we continue showing the cached data? (I mean, I'm aware only about the load more failure).

Oh, sorry, I meant that we (now) invalidate it when we do PTR in general, not only when the request fails. We display an error when requests fails.

> Yes, if you don't look at the big picture, I agree handling such cases is simple. It's similar with VM size - at the beginning one might feel like we can put everything there, since it's just 80 lines of code. However, if we continue applying this we end up with 500+ lines of code and the way back is very expensive.

> In this case, it might be one if statement, but if we handle 15 such cases, we have 15 if statements in a better case, in the worse case even more since the states affect each other. Every single if statement needs to be understood by whoever wants to change the class => it is often difficult to understand them when they fix random edge cases which often don't matter.

I am talking about the specific case where all these cases are handled, on the products list and its relativilty simple VM. As the code suggests, it handles all the cases to show either loading state or PTR... we don't need 15 cases for this.

malinajirka commented:
> Oh, sorry, I meant that we (now) invalidate it when we do PTR in general, not only when the request fails. We display an error when requests fails.

Ohh I see. The usual practice AFAIK is to clear the cache only when the request succeeds - right before we store the new data into the DB. Either case, since we don't have designs for `remote request failed but cached data available` state, I'd just go with the current solution which shows fullscreen error when PTR fails.
